File: src/view/gui.py
import tkinter as tk

from logic.board import Board
from logic.move import Move
from logic.pieces.piece import Colour, Piece
from logic.position import Position
from view.view import View

class GUI(View):
    def __init__(self) -> None:
        super().__init__()

        self.root = tk.Tk()
        self.root.title("Chess Board")

        self.tile_size = 80
        board_size = self.tile_size * 8

        self.canvas = tk.Canvas(self.root, width=board_size, height=board_size)
        self.canvas.pack()

        self.canvas.bind("<Button-1>", self._on_canvas_click)

    def _on_canvas_click(self, event):
        x, y = event.x // self.tile_size, event.y // self.tile_size
        y = 7 - y

        self._controller.on_tile_selected(x, y)

    def update_board(self, board: Board, selected_piece: Piece, legal_moves: list[Move]) -> None:
        self.canvas.delete("all")
        self._draw_chess_board(board, selected_piece, legal_moves)

    def _draw_chess_board(self, board, selected_piece = None, legal_moves = []):
        colours = ["#F0D9B5", "#B58863"]  # Light and dark squares

        for y in range(8):
            for x in range(8):
                colour = colours[(x + y) % 2]
                if selected_piece is not None:
                    possible_positions = [move.pos for move in legal_moves] 
                    if Position(x, 7-y) in possible_positions:
                        colour = "#ADD8E6"  # Highlight legal moves

                self.canvas.create_rectangle(
                    x * self.tile_size, 
                    y * self.tile_size, 
                    (x + 1) * self.tile_size, 
                    (y + 1) * self.tile_size, 
                    fill=colour,
                    outline=colour,
                )

                piece = board.piece_at(x, 7-y)

                if piece:
                    text_colour = "white" if piece.colour == Colour.WHITE else "black"
                    self.canvas.create_text(
                        (x + 0.5) * self.tile_size, 
                        (y + 0.5) * self.tile_size, 
                        text=piece.__class__.__name__[0], 
                        fill=text_colour, 
                        font=("Arial", 32, "bold")
                    )

                # Cell annotations
                text_colour = colours[(x + y + 1) % 2] # the other colour

                if x == 0: # numbers in the top left of the first column
                    self.canvas.create_text(
                        (x + 0.15) * self.tile_size, 
                        (y + 0.15) * self.tile_size, 
                        text=8-y, 
                        fill=text_colour, 
                        font=("Arial", 10, "bold")
                    )
                if y == 7: # numbers in the top left of the first column
                    self.canvas.create_text(
                        (x + 0.85) * self.tile_size, 
                        (y + 0.85) * self.tile_size, 
                        text="abcdefgh"[x], 
                        fill=text_colour, 
                        font=("Arial", 10, "bold")
                    )

    def show(self) -> None:
        self.root.mainloop()

File: cpp/src/view/gui.cpp
#include "gui.hpp"

#include "../model/utils/utils.hpp"

#include <SFML/Graphics/Color.hpp>
#include <SFML/System/Vector2.hpp>

GUI::GUI() {
    window.create(sf::VideoMode(WINDOW_SIZE, WINDOW_SIZE), "Chess Board");
    load_textures();

    font.loadFromFile("res/arial.ttf");
}

void GUI::update_board(
    const Board& b, int8_t selected_square, std::vector<Move> legal_moves
) {
    window.clear();
    draw_board(selected_square, legal_moves);
    draw_pieces(b);
    window.display();
}

void GUI::notify_stalemate() {}

void GUI::notify_checkmate() {}

void GUI::handle_events() {
    sf::Event event;
    while (window.pollEvent(event))
        if (event.type == sf::Event::Closed)
            window.close();
        else if (event.type == sf::Event::MouseButtonPressed)
            handle_click(event.mouseButton.x, event.mouseButton.y);
}

void GUI::load_textures() {
    const std::string names[6] =
        {"rook", "knight", "bishop", "queen", "king", "pawn"
        }; // don't touch the order, it's reflecting the one in the Piece enum
    for (int i = 0; i < 6; ++i) {
        textures[i][0].loadFromFile("res/pieces/white-" + names[i] + ".png");
        textures[i][1].loadFromFile("res/pieces/black-" + names[i] + ".png");
    }
}

void GUI::handle_click(int x, int y) {
    int file = x / TILE_SIZE;
    int rank = 7 - (y / TILE_SIZE);
    controller->on_tile_selected(file, rank);
}

void GUI::draw_annotation(int file, int rank) {
    if (file == 0) {
        sf::Text annotation(std::to_string(rank + 1), font);
        annotation.setStyle(sf::Text::Bold);
        annotation.setCharacterSize(16);
        annotation.setFillColor(rank % 2 == 0 ? colours[1] : colours[0]);
        annotation.setPosition(
            (file + .05) * TILE_SIZE,
            (7 - rank + .05) * TILE_SIZE
        );
        window.draw(annotation);
    }

    if (rank == 0) {
        sf::Text annotation("abcdefgh"[file], font);
        annotation.setCharacterSize(16);
        annotation.setOrigin(16, 16);
        annotation.setStyle(sf::Text::Bold);
        annotation.setFillColor(file % 2 == 0 ? colours[1] : colours[0]);
        annotation.setPosition(
            (file + 1) * TILE_SIZE,
            (7 - rank + .95) * TILE_SIZE
        );
        window.draw(annotation);
    }
}

void GUI::draw_board(int selected_square, std::vector<Move> moves) {
    sf::RectangleShape square(sf::Vector2f(TILE_SIZE, TILE_SIZE));
    for (int rank = 0; rank < 8; ++rank) {
        for (int file = 0; file < 8; ++file) {
            int8_t index = Coords{file, rank}.to_index();
            square.setPosition(file * TILE_SIZE, (7 - rank) * TILE_SIZE);
            if (index == selected_square)
                square.setFillColor(
                    (file + rank) % 2 == 0 ? alt_colours[0] : alt_colours[1]
                );
            else
                square.setFillColor(
                    (file + rank) % 2 == 0 ? colours[0] : colours[1]
                );
            window.draw(square);
            draw_annotation(file, rank);

            std::vector<int8_t> targets = to_target_square(moves);
            if (std::find(targets.begin(), targets.end(), index)
                != targets.end()) {
                float r = .15 * TILE_SIZE;
                sf::CircleShape circle{r};
                sf::Color c(0x00000055);
                circle.setFillColor(c);
                circle.setOrigin(r, r);
                circle.setPosition(
                    (file + .5) * TILE_SIZE,
                    (7 - rank + .5) * TILE_SIZE
                );
                window.draw(circle);
            }
        }
    }
}

void GUI::draw_pieces(const Board& board) {
    for (int i = 0; i < 64; ++i) {
        int piece = board.piece_at(i);
        if (piece != Piece::None) {
            int colour = board.colour_at(i) == Colour::White ? 0 : 1;
            pieces[i].setTexture(textures[piece - 1][colour]);

            sf::Vector2 center = textures[piece - 1][colour].getSize() / 2u;
            pieces[i].setOrigin(center.x, center.y);

            pieces[i].setPosition(
                (i % 8 + .5) * TILE_SIZE,
                (7 - (int) (i / 8) + .5) * TILE_SIZE
            );
            window.draw(pieces[i]);
        }
    }
}

void GUI::show() {
    while (window.isOpen())
        handle_events();
}
